src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,970 @@
+        "encoder_padding_mask",  # B x T
+        "encoder_embedding",  # B x T x C
+        "encoder_states",  # List[T x B x C]
+        "encoder_attn",
+    ],
+)
+
+
+class BartEncoder(nn.Module):
+    """
+    Transformer encoder consisting of *config.encoder_layers* layers. Each layer
+    is a :class:`EncoderLayer`.
+
+    Args:
+        config (argparse.Namespace): parsed command-line arguments
+        dictionary (~fairseq.data.Dictionary): encoding dictionary
+        embed_tokens (torch.nn.Embedding): input embedding
+    """
+
+    def __init__(self, config: BartConfig, embed_tokens):
+        super().__init__()
+        self.register_buffer("version", torch.Tensor([3]))
+
+        self.dropout = config.dropout
+        self.layerdrop = config.encoder_layerdrop
+        self.output_attentions = config.output_attentions
+        self.output_hidden_states = config.output_hidden_states
+
+        embed_dim = embed_tokens.embedding_dim
+        self.padding_idx = embed_tokens.padding_idx
+        self.max_source_positions = config.max_position_embeddings
+
+        self.embed_tokens = embed_tokens
+
+        self.embed_positions = LearnedPositionalEmbedding(config.max_position_embeddings, embed_dim, self.padding_idx,)
+        self.layers = nn.ModuleList([EncoderLayer(config) for _ in range(config.encoder_layers)])
+        self.layernorm_embedding = LayerNorm(embed_dim)
+
+    def forward_embedding(self, input_ids):
+        # embed tokens and positions
+        embedded_tokens = self.embed_tokens(input_ids)
+        x = embedded_tokens + self.embed_positions(input_ids)
+        x = self.layernorm_embedding(x)
+        x = F.dropout(x, p=self.dropout, training=self.training)
+        return x, embedded_tokens
+
+    def forward(self, input_ids, **unused):  # TODO(SS): this will need more
+        """
+        Args:
+            input_ids (LongTensor): tokens in the source language of shape
+                `(batch, src_len)`
+            src_lengths (torch.LongTensor): lengths of each source sentence of
+                shape `(batch)`
+        Returns:
+            namedtuple:
+                - **encoder_out** (Tensor): the last encoder layer's output of
+                  shape `(src_len, batch, embed_dim)`
+                - **encoder_padding_mask** (ByteTensor): the positions of
+                  padding elements of shape `(batch, src_len)`
+                - **encoder_embedding** (Tensor): the (scaled) embedding lookup
+                  of shape `(batch, src_len, embed_dim)`
+                - **encoder_states** (List[Tensor]): all intermediate
+                  hidden states of shape `(src_len, batch, embed_dim)`.
+                  Only populated if *return_all_hiddens* is True.
+        """
+        x, encoder_embedding = self.forward_embedding(input_ids)
+
+        # B x T x C -> T x B x C
+        x = x.transpose(0, 1)
+
+        # compute padding mask
+        encoder_padding_mask = input_ids.eq(self.padding_idx)
```

Comment:
Concerning 2) I think it's important we have the same API as the other models

Reply:
The way the fairseq code works is that there are two attention masks created in the forward pass for the decoder:
the padding_mask is shaped (bs, seq_len) and causes the model to ignore pad tokens. [1] 
the attention_mask is shaped (src_len, tgt_len) and is used in the decoder to prevent it from attending to future tokens in the target sequence. [2]
#### Plan to consolidate masks:

* force the caller to combine to one attention_mask of shape (bsz, tgt_len, src_len), and achieve API consistency at the cost of a small amount of memory.
* add `def prepare_input_masks_for_masked_lm(input_ids, decoder_input_ids=None, encoder_hidden_states=None)` utility makes masks for callers.

[1] seq_len can be different between encoder and decoder in seq2seq tasks.
[2] A major point of Bart is that the decoder is left-to-right and the encoder is bidirectional. In T5 and UniLM this technique is called prefix LM/the mask is called "causal mask with prefix".